pwndbg/commands/cyclic.py

```diff
@@ -0,0 +1,88 @@
+import argparse
+import string
+
+import gdb
+from pwnlib.util.cyclic import cyclic
+from pwnlib.util.cyclic import cyclic_find
+
+import pwndbg.commands
+import pwndbg.gdblib.arch
+from pwndbg.color import message
+
+parser = argparse.ArgumentParser(description="Cyclic pattern creator/finder")
+
+parser.add_argument(
+    "-a",
+    "--alphabet",
+    metavar="charset",
+    default=string.ascii_lowercase,
+    type=str.encode,
+    help="The alphabet to use in the cyclic pattern",
+)
+
+parser.add_argument(
+    "-n",
+    "--length",
+    metavar="length",
+    type=int,
+    help="Size of the unique subsequences (defaults to the pointer size for the current arch)",
+)
+
+group = parser.add_mutually_exclusive_group(required=False)
+group.add_argument(
+    "-l",
+    "-o",
+    "--offset",
+    "--lookup",
+    dest="lookup",
+    metavar="lookup_value",
+    type=str,
+    help="Do a lookup instead of printing the sequence (accepts constant values as well as expressions)",
+)
+
+group.add_argument(
+    "count",
+    type=int,
+    nargs="?",
+    default=None,
+    help="Number of characters to print from the sequence (default: print the entire sequence)",
+)
+
+
+@pwndbg.commands.ArgparsedCommand(parser, command_name="cyclic")
+def cyclic_cmd(alphabet, length, lookup, count):
+    if length:
+        # Convert from gdb.Value
+        length = int(length)
+    else:
+        # Placing the default here as placing it in the arguments
+        # will default to 4 for some reason
+        length = pwndbg.gdblib.arch.ptrsize
```

Comment:
The exact reason for this is that expressions used for default arguments are evaluated right away in the moment when the function is created. You can see this here:
<img width="705" alt="image" src="https://user-images.githubusercontent.com/10009354/198156000-96916f51-406b-4d8f-aecc-59746d92d671.png">

And a text form:
```
In [1]: import dis

In [2]: dis.dis('def foo(x=some_func()): pass')
  1           0 LOAD_NAME                0 (some_func)
              2 CALL_FUNCTION            0
              4 BUILD_TUPLE              1
              6 LOAD_CONST               0 (<code object foo at 0x1101c6d40, file "<dis>", line 1>)
              8 LOAD_CONST               1 ('foo')
             10 MAKE_FUNCTION            1 (defaults)
             12 STORE_NAME               1 (foo)
             14 LOAD_CONST               2 (None)
             16 RETURN_VALUE

Disassembly of <code object foo at 0x1101c6d40, file "<dis>", line 1>:
  1           0 LOAD_CONST               0 (None)
              2 RETURN_VALUE
```

As you can see, during the execution of code that defines the function, the default argument is evaluated and so the `some_func()` is called and its result is stored into a tuple with 1 elements (`BUILD_TUPLE 1`), then the actual code of the function is loaded and the function is created with its default arguments and then it is stored in `foo` name.

This is often called a "mutable defaults arguments trap". See e.g.:
```py
In [12]: def bar(x, y=[]):
    ...:     y.append(x)
    ...:     return y
    ...:

In [13]: bar(1)
Out[13]: [1]

In [14]: bar(2)
Out[14]: [1, 2]

In [15]: bar(3)
Out[15]: [1, 2, 3]
```

Reply:
Can we change the comment since we know the exact reason?

Reply:
The comment was more about the fact it resolves to `4` rather than why it's set to a static value since I couldn't find that out, but I see the point. Will change that.

Reply:
It's because in init.py we load the commands before gdblib, so the arch is not set until later.

This was unintentional, I'll check if I can flip the order without breaking things (the current situation is at least better than the old one, where all imports were alphabetical so commands would get initialized in the middle of other modules being initialized).